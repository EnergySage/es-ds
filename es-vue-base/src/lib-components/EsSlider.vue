<template>
    <client-only>
        <vue-slider
            v-model="sliderValue"
            class="es-slider"
            :data="data"
            :marks="marks"
            :min="min"
            :max="max"
            :dot-size="28"
            :dot-attrs="{ 'aria-label': ariaLabel }"
            :tooltip="tooltip"
            lazy
            v-bind="$attrs"
            :contained="true"
            v-on="$listeners">
            <!-- Tooltip above the slider thumb -->
            <template #tooltip="{ value, focus }">
                <div
                    class="slider-tooltip d-flex align-items-center text-white justify-content-center m-0 bg-blue-600 font-weight-boldest"
                    :class="[{ focus }]">
                    {{ tooltipFormatter(value) }}
                </div>
            </template>
            <!-- Labels below the process bar -->
            <template #label="{ label, active }">
                <div :class="['slider-label', 'vue-slider-mark-label', { active }]">
                    {{ labelFormatter(label) }}
                </div>
            </template>
        </vue-slider>
    </client-only>
</template>

<script lang="js">
import vueSlider from 'vue-slider-component/dist-css/vue-slider-component.common';

export default {
    name: 'EsSlider',
    components: {
        vueSlider,
    },
    props: {
        /**
         * Starting value of slider thumb
         */
        startingValue: {
            type: Number,
            required: true,
        },
        /**
         * Array that contains valid steps
         */
        data: {
            type: Array,
            required: true,
        },
        /**
         * Function that chooses where to add marks to the track
         */
        marks: {
            type: [Array, Function],
            required: true,
        },
        /**
         * Aria Label for Dot
         */
        ariaLabel: {
            type: String,
            required: false,
            default: 'Select a number',
        },
        /**
         * Function that modifies label value
         */
        labelFormatter: {
            type: Function,
            required: false,
            default: (val) => val,
        },
        /**
         * Function that modifies tooltip value
         */
        tooltipFormatter: {
            type: Function,
            required: false,
            default: (val) => val,
        },
        /**
         * Choose when tooltip displays
         */
        tooltip: {
            type: String,
            required: false,
            default: 'always',
        },
    },
    data() {
        return {
            sliderValue: this.startingValue,
            min: this.data[0],
            max: this.data[this.data.length - 1],
        };
    },
    watch: {
        startingValue: {
            // Watch the prop and update the value to avoid mutation
            handler(newVal) {
                this.sliderValue = newVal > 0 ? newVal : this.startingValue;
            },
        },
    },
};
</script>
