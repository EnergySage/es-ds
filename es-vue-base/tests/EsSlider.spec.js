import { mount } from '@vue/test-utils';
import EsSlider from '@/src/lib-components/EsSlider.vue';
import jestVue from '@/tests/jest.vue.config';

describe('EsSlider', () => {
    // Basic test; should exist for most components
    test(`<EsSlider
            :startingValue="100"
            :data="[0,100,500]"
            :marks="[0,500]"
            :min="100"
            :max="500"
            ariaLabel="Select a number" />`, async () => {
        const wrapper = mount(EsSlider, {
            ...jestVue,
            propsData: {
                startingValue: 300,
                data: [100, 300, 500],
                marks: () => [100, 500],
                min: 100,
                max: 500,
                ariaLabel: 'Select a number',
            },
        });
        const a11y = await axe(wrapper.element);

        expect(wrapper.vm).toBeTruthy();
        expect(wrapper.html()).toMatchSnapshot();
        expect(a11y).toHaveNoViolations();
    });

    test('<EsSlider :startingValue="50" />', async () => {
        const wrapper = mount(EsSlider, {
            ...jestVue,
            propsData: {
                startingValue: 50,
                data: [],
                marks: () => [],
                min: 0,
                max: 100,
                ariaLabel: 'Select a number',
            },
        });

        expect(wrapper.html()).toMatchSnapshot();

        const tooltip = wrapper.find('div.slider-tooltip');
        expect(tooltip.text()).toContain('50');
        expect(wrapper.vm.sliderValue).toBe(50);
    });

    test('<EsSlider variant="secondary" />', async () => {
        const wrapper = mount(EsSlider, {
            ...jestVue,
            propsData: {
                startingValue: 50,
                data: [],
                marks: () => [],
                min: 0,
                max: 100,
                ariaLabel: 'Select a number',
                variant: 'secondary',
            },
        });
        expect(wrapper.html()).toMatchSnapshot();
    });

    test('<EsSlider /> tooltipFormatter works', async () => {
        const wrapper = mount(EsSlider, {
            ...jestVue,
            propsData: {
                startingValue: 50,
                data: [],
                marks: () => [],
                min: 0,
                max: 100,
                ariaLabel: 'Select a number',
                tooltipFormatter: (v) => `$${v}`,
            },
        });

        expect(wrapper.html()).toMatchSnapshot();

        const tooltip = wrapper.find('div.slider-tooltip');
        expect(tooltip.text()).toContain('$');
    });
});
