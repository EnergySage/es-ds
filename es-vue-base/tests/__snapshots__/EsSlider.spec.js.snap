// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EsSlider <EsSlider
            :startingValue="100"
            :data="[0,100,500]"
            :marks="[0,500]"
            :min="100"
            :max="500"
            ariaLabel="Select a number" /> 1`] = `
"<div min="100" max="500">
  <div class="es-slider vue-slider vue-slider-ltr" style="padding: 14px 14px; width: auto; height: 4px;">
    <div class="vue-slider-rail">
      <div class="vue-slider-process" style="height: 100%; top: 0px; left: 0%; width: 50%; transition-property: width,left; transition-duration: 0.5s;"></div>
      <div class="vue-slider-marks">
        <div class="vue-slider-mark vue-slider-mark-active" style="height: 100%; width: 4px; left: 0%;">
          <div class="vue-slider-mark-step vue-slider-mark-step-active"></div>
          <div class="slider-label vue-slider-mark-label active">
            100
          </div>
        </div>
        <div class="vue-slider-mark" style="height: 100%; width: 4px; left: 100%;">
          <div class="vue-slider-mark-step"></div>
          <div class="slider-label vue-slider-mark-label">
            500
          </div>
        </div>
      </div>
      <div aria-valuetext="300" class="vue-slider-dot" role="slider" aria-valuenow="300" aria-valuemin="100" aria-valuemax="500" aria-orientation="horizontal" tabindex="0" aria-label="Select a number" style="width: 28px; height: 28px; transform: translate(-50%, -50%); top: 50%; left: 50%; transition: left 0.5s;">
        <div class="vue-slider-dot-handle"></div>
        <div class="vue-slider-dot-tooltip vue-slider-dot-tooltip-top vue-slider-dot-tooltip-show">
          <div class="slider-tooltip d-flex h5 align-items-center text-white justify-content-center m-0 bg-primary">
            300
          </div>
        </div>
      </div>
    </div>
  </div>
</div>"
`;

exports[`EsSlider <EsSlider /> tooltipFormatter works 1`] = `
"<div min="0" max="100">
  <div class="es-slider vue-slider vue-slider-ltr" style="padding: 14px 14px; width: auto; height: 4px;">
    <div class="vue-slider-rail">
      <div class="vue-slider-process" style="height: 100%; top: 0px; left: 0%; width: 100%; transition-property: width,left; transition-duration: 0.5s;"></div>
      <div class="vue-slider-marks"></div>
      <div aria-valuetext="50" class="vue-slider-dot" role="slider" aria-valuenow="50" aria-valuemin="0" aria-valuemax="100" aria-orientation="horizontal" tabindex="0" aria-label="Select a number" style="width: 28px; height: 28px; transform: translate(-50%, -50%); top: 50%; left: 100%; transition: left 0.5s;">
        <div class="vue-slider-dot-handle"></div>
        <div class="vue-slider-dot-tooltip vue-slider-dot-tooltip-top vue-slider-dot-tooltip-show">
          <div class="slider-tooltip d-flex h5 align-items-center text-white justify-content-center m-0 bg-primary">
            $50
          </div>
        </div>
      </div>
    </div>
  </div>
</div>"
`;

exports[`EsSlider <EsSlider :startingValue="50" /> 1`] = `
"<div min="0" max="100">
  <div class="es-slider vue-slider vue-slider-ltr" style="padding: 14px 14px; width: auto; height: 4px;">
    <div class="vue-slider-rail">
      <div class="vue-slider-process" style="height: 100%; top: 0px; left: 0%; width: 100%; transition-property: width,left; transition-duration: 0.5s;"></div>
      <div class="vue-slider-marks"></div>
      <div aria-valuetext="50" class="vue-slider-dot" role="slider" aria-valuenow="50" aria-valuemin="0" aria-valuemax="100" aria-orientation="horizontal" tabindex="0" aria-label="Select a number" style="width: 28px; height: 28px; transform: translate(-50%, -50%); top: 50%; left: 100%; transition: left 0.5s;">
        <div class="vue-slider-dot-handle"></div>
        <div class="vue-slider-dot-tooltip vue-slider-dot-tooltip-top vue-slider-dot-tooltip-show">
          <div class="slider-tooltip d-flex h5 align-items-center text-white justify-content-center m-0 bg-primary">
            50
          </div>
        </div>
      </div>
    </div>
  </div>
</div>"
`;

exports[`EsSlider <EsSlider variant="secondary" /> 1`] = `
"<div min="0" max="100" variant="secondary">
  <div class="es-slider vue-slider vue-slider-ltr" style="padding: 14px 14px; width: auto; height: 4px;" variant="secondary">
    <div class="vue-slider-rail">
      <div class="vue-slider-process" style="height: 100%; top: 0px; left: 0%; width: 100%; transition-property: width,left; transition-duration: 0.5s;"></div>
      <div class="vue-slider-marks"></div>
      <div aria-valuetext="50" class="vue-slider-dot" role="slider" aria-valuenow="50" aria-valuemin="0" aria-valuemax="100" aria-orientation="horizontal" tabindex="0" aria-label="Select a number" style="width: 28px; height: 28px; transform: translate(-50%, -50%); top: 50%; left: 100%; transition: left 0.5s;">
        <div class="vue-slider-dot-handle"></div>
        <div class="vue-slider-dot-tooltip vue-slider-dot-tooltip-top vue-slider-dot-tooltip-show">
          <div class="slider-tooltip d-flex h5 align-items-center text-white justify-content-center m-0 bg-primary">
            50
          </div>
        </div>
      </div>
    </div>
  </div>
</div>"
`;
